src/internal_events/open.rs

```diff
@@ -0,0 +1,129 @@
+use super::InternalEvent;
+use metrics::gauge;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+pub type OpenTokenDyn = OpenToken<Box<dyn Fn(usize) + 'static + Send>>;
+
+#[derive(Debug)]
+pub struct ConnectionOpen {
+    pub count: usize,
+}
+
+impl InternalEvent for ConnectionOpen {
+    fn emit_metrics(&self) {
+        gauge!("open_connections", self.count as f64);
+    }
+}
+
+#[derive(Clone)]
+pub struct OpenGauge {
+    gauge: Arc<AtomicUsize>,
+}
+
+impl OpenGauge {
+    pub fn new() -> Self {
+        OpenGauge {
+            gauge: Arc::default(),
+        }
+    }
+
+    /// Increments and emits value once created.
+    /// Decrements and emits value once dropped.
+    pub fn open<E: Fn(usize)>(self, emitter: E) -> OpenToken<E> {
+        gauge_add(&self.gauge, 1, &emitter);
+        OpenToken {
+            gauge: self.gauge,
+            emitter,
+        }
+    }
+}
+
+impl Default for OpenGauge {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct OpenToken<E: Fn(usize)> {
+    gauge: Arc<AtomicUsize>,
+    emitter: E,
+}
+
+impl<E: Fn(usize)> Drop for OpenToken<E> {
+    fn drop(&mut self) {
+        gauge_add(&self.gauge, -1, &self.emitter);
+    }
+}
+
+/// If reporting gauges from multiple threads, they can end up in a wrong order
+/// resulting in having wrong value for a prolonged period of time.
+/// This function performs a synchronization procedure that corrects that.
+fn gauge_add(gauge: &AtomicUsize, add: isize, emitter: impl Fn(usize)) {
+    // Lock-free procedure with eventual consistency
+
+    // Initialize value and acquire older writes to gauge metric.
+    let mut value = gauge.load(Ordering::Acquire);
```

Comment:
It looks like this is doing https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html#method.fetch_add ?

Reply:
Not really, `fetch_add` would be ideal if there wasn't the issue of unordered `gauge!(...)` calls. For example, with `fetch_add`, two threads, both trying to increment gauge by 1, and `Atomic` `gauge` with start value of 0, the following situation can happen:
- thread 0: v0 = `gauge.fetch_add(1)` = 0; --------------`gauge = 1`
- thread 1: v1 = `gauge.fetch_add(1)` = 1; --------------`gauge = 2`
- thread 1: `gauge!(open_connections = v1 + 1)`; ----gauge `open_connection` = 2
- thread 0: `gauge!(open_connections = v0 + 1)`; ----gauge `open_connection` = 1

In this scenario `open_connection` gauge would be reported as 1 for as long as nothing else updates it. So to avoid that we use the commented algorithm.

Reply:
Ah ok, got you. :)